### PR TITLE
feat: [PLG-1984]: Support ff tiers

### PR DESCRIPTION
**Pull request by mtran7**
## Describe your changes
AWS API Identifiers follow the format `<Module>_<Plan>_<Premier (optional)>_<Quantity>`. So we need to parse the quantity of users, plan, number of Client MAUs (specifically for FF) correctly.
For Plan, we should support shorthand, "TM" for Team and "EN" for Enterprise due to character limit on AWS side for other modules
image

All is scoped to Feature Flags and needs to be verified on QA
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45798)

### CODE REVIEW

**Review by neelam-harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45798)* status: 2 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java` line 97 at r1](https://reviewable.io/reviews/harness/harness-core/45798#-NRertMv0zThCouTflLe:-NRertMv0zThCouTflLf:b-5ajhdm) ([raw file](https://github.com/harness/harness-core/blob/e8660443dd150b7efdc53d342fdad75a6d19acb0/400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java#L97)):*
> ```Java
> public class LicenseServiceImpl implements LicenseService {
>   private static final Long TEST_FF_NUMBER_OF_CLIENT_MAUS = 1000000L;
>   private static final String FREE_CLIENT_MAUS = "25K";
> ```

Is this supposed to be a `String` or `Integer` since it's representing a number value ?
___
*[`400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java` line 98 at r1](https://reviewable.io/reviews/harness/harness-core/45798#-NRes0kwAMthlew1Gd-s:-NRes0kwAMthlew1Gd-t:bum9rfu) ([raw file](https://github.com/harness/harness-core/blob/e8660443dd150b7efdc53d342fdad75a6d19acb0/400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java#L98)):*
> ```Java
>   private static final Long TEST_FF_NUMBER_OF_CLIENT_MAUS = 1000000L;
>   private static final String FREE_CLIENT_MAUS = "25K";
>   private static final String TEAM_CLIENT_MAUS = "100K";
> ```

Same for this one too.
___
*[`400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java` line 289 at r1](https://reviewable.io/reviews/harness/harness-core/45798#-NResRbu7jeMlChECxXP:-NResRbu7jeMlChECxXQ:b69f63c) ([raw file](https://github.com/harness/harness-core/blob/e8660443dd150b7efdc53d342fdad75a6d19acb0/400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java#L289)):*
> ```Java
>   private String getDimensionModule(String dimension) {
>     String module = "";
>     if (dimension) {
> ```

What do we mean by `if (dimension)` here ? Is `dimension` a true or false value ?
___
*[`400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java` line 304 at r1](https://reviewable.io/reviews/harness/harness-core/45798#-NReshg49-Y5bZugKzBU:-NReshg49-Y5bZugKzBV:bmfbrow) ([raw file](https://github.com/harness/harness-core/blob/e8660443dd150b7efdc53d342fdad75a6d19acb0/400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java#L304)):*
> ```Java
> 
>     // Handle K (1000) and M (1000000) units
>     if (tempQuantity.contains("K")) {
> ```

I don't think we need to do this string manipulation here. Can we not declare the variables appropriately to begin with ? 



_Code quote:_
```Java
    // Handle K (1000) and M (1000000) units
    if (tempQuantity.contains("K")) {
      tempQuantity = tempQuantity.replace("K", "000");
    } else if (tempQuantity.contains("M")) {
      tempQuantity = tempQuantity.replace("M", "000000");
    }
```

File: 400-rest/src/test/java/software/wings/licensing/LicenseServiceImplTest.java [hunk 1]
```diff
@@ -103,9 +103,9 @@ public void shouldGetExpectedPlansFromDimensions() {
   @Owner(developers = MTRAN7)
   @Category(UnitTests.class)
   public void shouldGetExpectedNumberOfClientMAUsFromPlan() {
-    private static final Long FREE_CLIENT_MAUS = (long) 25000;
-    private static final Long TEAM_CLIENT_MAUS = (long) 100000;
-    private static final Long ENTERPRISE_CLIENT_MAUS = (long) 1000000;
+    Long FREE_CLIENT_MAUS = (long) 25000;
```
Comment (neelam-harness) [inline 400-rest/src/test/java/software/wings/licensing/LicenseServiceImplTest.java:106]: instead of `(long) 25000` you can declare it as `25000L`

**Review by neelam-harness — CHANGES_REQUESTED** [review]

File: 400-rest/src/test/java/software/wings/licensing/LicenseServiceImplTest.java (hunk 1 above)
Comment (mtran7) [reply 400-rest/src/test/java/software/wings/licensing/LicenseServiceImplTest.java:107]: Ok will keep this in mind for next pr update. Thanks!

Comment (neelam-harness) [reply 400-rest/src/test/java/software/wings/licensing/LicenseServiceImplTest.java:107]: it's a simple change and could have been included in this pr itself :) 
But don't want to block you from merging. Please update it when you plan to update this. Thanks

**Review by neelam-harness — APPROVED** [review]

**Review by neelam-harness — APPROVED** [review]

**Review by neelam-harness — DISMISSED** [review]
:lgtm:

> Reviewed 5 of 6 files at r2, 1 of 3 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45798#-:-NRuxI8vC-IK_OdfflwE:bnfp4nl)* status: 8 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by mtran7 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45798)* status: 8 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`400-rest/src/test/java/software/wings/licensing/LicenseServiceImplTest.java` line 106 at r3](https://reviewable.io/reviews/harness/harness-core/45798#-NRtnQ6jlZ5VmHFqUFHI-r3-106:-NRxf1Sd4BK3R16NQJ3s:b43574s) ([raw file](https://github.com/harness/harness-core/blob/fef5700606f6b21e85e45312946533963e270845/400-rest/src/test/java/software/wings/licensing/LicenseServiceImplTest.java#L106)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

it's a simple change and could have been included in this pr itself :) 
But don't want to block you from merging. Please update it when you plan to update this. Thanks
</blockquote></details>

Made the change!

**Review by Xin-Harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45798)* status: 6 file(s) to review. 1 change requests. 2. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java` line 694 at r7](https://reviewable.io/reviews/harness/harness-core/45798#-NS3heY63wKUEjJMr5we:-NS3heY63wKUEjJMr5wf:ble5w1b) ([raw file](https://github.com/harness/harness-core/blob/19ea7655fe152cc4ab18ad7812fee00cb9e67a11/400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java#L694)):*
> ```Java
>     if (dimension instanceof java.lang.String) {
>       String[] result = dimension.split("_");
>       String tempPlan = result[1];
> ```

1.It could cause index out of bound exception if dimension doesn't have index 1. 
2. dimension is already a string, it doesn't need to check if it is an instance of java.lang.string
___
*[`400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java` line 708 at r7](https://reviewable.io/reviews/harness/harness-core/45798#-NS3mLpW2tflfU2V1QGN:-NS3mLpW2tflfU2V1QGO:bkldpih) ([raw file](https://github.com/harness/harness-core/blob/19ea7655fe152cc4ab18ad7812fee00cb9e67a11/400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java#L708)):*
> ```Java
>   // Gets Number of Client MAUs from plan
>   public Long getNumberOfClientMAUs(Edition plan) {
>     Long numberOfClientMAUs = FREE_CLIENT_MAUS;
> ```

Question here: Since the default number of client MAUs is FREE_CLIENT_MAUS, if the Edition is FREE itself, it will fall into else statement "Failed to get number of Client MAUs" instead of return default free client maus . Is this what it suppose to be? Any reason for the design.
___
*[`400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java` line 723 at r7](https://reviewable.io/reviews/harness/harness-core/45798#-NS3pZNL76vD-rTqCxVo:-NS3pZNL76vD-rTqCxVp:bknyw1z) ([raw file](https://github.com/harness/harness-core/blob/19ea7655fe152cc4ab18ad7812fee00cb9e67a11/400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java#L723)):*
> ```Java
>   public LicenseType getModuleLicenseType(String dimension, Edition plan) {
>     LicenseType licenseType = null;
>     String[] result = dimension.split("_");
> ```

It need to make sure dimension is not null, then to split the string, otherwise it will cause null pointer exception.
___
*[`400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java` line 289 at r1](https://reviewable.io/reviews/harness/harness-core/45798#-NResRbu7jeMlChECxXP:-NS3rJa99_GlJghWcAre:benpsw9) ([raw file](https://github.com/harness/harness-core/blob/e8660443dd150b7efdc53d342fdad75a6d19acb0/400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java#L289)):*
<details><summary><i>Previously, mtran7 wrote…</i></summary><blockquote>

It should be a string value. I've updated to dimension instanceof java.lang.String
</blockquote></details>

Same comment as above.
___
*[`400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java` line 152 at r7](https://reviewable.io/reviews/harness/harness-core/45798#-NS3q7XnETdNdvh0fGMq:-NS3q7XnETdNdvh0fGMr:b-q7o1pt) ([raw file](https://github.com/harness/harness-core/blob/19ea7655fe152cc4ab18ad7812fee00cb9e67a11/400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java#L152)):*
> ```Java
>     log.info("Order Quantity=[{}]", orderQuantity);
> 
>     // Replace orderQuantity with quantity provided in dimension for FF module
> ```

The code is self explanatory , would recommend not use comment for most of the time, due to code convension.

**Review by mtran7 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45798)* status: 6 file(s) to review. 1 change requests. 5. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java` line 694 at r7](https://reviewable.io/reviews/harness/harness-core/45798#-NS3heY63wKUEjJMr5we:-NS7R8iP2XW74Ijo3vnI:b-20dg5c) ([raw file](https://github.com/harness/harness-core/blob/19ea7655fe152cc4ab18ad7812fee00cb9e67a11/400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java#L694)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

1.It could cause index out of bound exception if dimension doesn't have index 1. 
2. dimension is already a string, it doesn't need to check if it is an instance of java.lang.string
</blockquote></details>

Added check for #1
For #2, it is expected to be a string, but in the slim chance it is not, I don't want to try and use the split method on an undefined or null variable. Please advise if there is a better way to handle for this scenario. I've used this instance of check in other areas as well.



_Code quote:_
```Java
dimension instanceof java.lang.String)
```
___
*[`400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java` line 723 at r7](https://reviewable.io/reviews/harness/harness-core/45798#-NS3pZNL76vD-rTqCxVo:-NS7S7oI15G9n7Gnn-Sg:b-oz8o77) ([raw file](https://github.com/harness/harness-core/blob/19ea7655fe152cc4ab18ad7812fee00cb9e67a11/400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java#L723)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

It need to make sure dimension is not null, then to split the string, otherwise it will cause null pointer exception.
</blockquote></details>

Good catch
___
*[`400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java` line 289 at r1](https://reviewable.io/reviews/harness/harness-core/45798#-NResRbu7jeMlChECxXP:-NS7V07v052t5z47t_m6:br9el56) ([raw file](https://github.com/harness/harness-core/blob/e8660443dd150b7efdc53d342fdad75a6d19acb0/400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java#L289)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

Same comment as above.
</blockquote></details>

I don't want to try and use the split method on an undefined or null variable. Please let me know if there's a better way to handle.
___
*[`400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java` line 152 at r7](https://reviewable.io/reviews/harness/harness-core/45798#-NS3q7XnETdNdvh0fGMq:-NS7SFboCjfuciqE9YCz:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/19ea7655fe152cc4ab18ad7812fee00cb9e67a11/400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java#L152)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

The code is self explanatory , would recommend not use comment for most of the time, due to code convension.
</blockquote></details>

Done.

**Review by Xin-Harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45798)* status: 6 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java` line 694 at r7](https://reviewable.io/reviews/harness/harness-core/45798#-NS3heY63wKUEjJMr5we:-NS7spXG9z3-tw_ld41B:b-a4cai) ([raw file](https://github.com/harness/harness-core/blob/19ea7655fe152cc4ab18ad7812fee00cb9e67a11/400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java#L694)):*
<details><summary><i>Previously, mtran7 wrote…</i></summary><blockquote>

Added check for #1
For #2, it is expected to be a string, but in the slim chance it is not, I don't want to try and use the split method on an undefined or null variable. Please advise if there is a better way to handle for this scenario. I've used this instance of check in other areas as well.
</blockquote></details>

For #1, 
since array index start from 0, when length is 1, only result[0] has content, only if length >=2, result[1] would have the content. 
Here is the right way to check
      if (result.length < 2) {
        log.error("Failed to get plan for dimension:[{}]", dimension);
        return plan;
      }
for #2 if the method is taking a string type and being successfully called, the dimension will be a string type for sure and can't be other type, this is how java make sure the matching method has being called. If dimension is not a string type, java would not allow call getDimensionPlan and pass other types. The only thing you need to check is dimension can't be null, otherwise it will raise null pointer exception to call dimensio.split.

 if (dimension!= null) {
   String[] result = dimension.split("_");
}

**Review by Xin-Harness — APPROVED** [review]
> Reviewed 2 of 6 files at r2, 1 of 2 files at r7, 1 of 3 files at r8, 1 of 2 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45798)* status: 3 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by neelam-harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45798)* status: 3 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java` line 692 at r10](https://reviewable.io/reviews/harness/harness-core/45798#-NS84nmiBkPCXRVc0zgK:-NS84nmiBkPCXRVc0zgL:bnea1vp) ([raw file](https://github.com/harness/harness-core/blob/16e5652812f1c2eb325818a47a8912c1bdce547f/400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java#L692)):*
> ```Java
>   public Edition getDimensionPlan(String dimension) {
>     Edition plan = null;
>     if (dimension != null) {
> ```

Why don't you use the `StringUtils.isNotBlank(dimension)` and then you can get rid of the following `if (result.length < 2) {` condition as well?
___
*[`400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java` line 731 at r10](https://reviewable.io/reviews/harness/harness-core/45798#-NS87-aR5RK6-emX4n_6:-NS87-aR5RK6-emX4n_7:bbb1ek0) ([raw file](https://github.com/harness/harness-core/blob/16e5652812f1c2eb325818a47a8912c1bdce547f/400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java#L731)):*
> ```Java
>       String[] result = dimension.split("_");
> 
>       if (result[0].equals("FF") && (plan.equals(Edition.TEAM) || plan.equals(Edition.ENTERPRISE))) {
> ```

Make  "FF" as a constant. 
`private static String FEATURE_FLAGS = "FF";`



_Code quote:_
```Java
String[] result = dimension.split("_");
```
___
*[`400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java` line 289 at r1](https://reviewable.io/reviews/harness/harness-core/45798#-NResRbu7jeMlChECxXP:-NS87aO1EQ5GPCrjfTEI:b-je45fh) ([raw file](https://github.com/harness/harness-core/blob/e8660443dd150b7efdc53d342fdad75a6d19acb0/400-rest/src/main/java/software/wings/service/impl/AwsMarketPlaceApiHandlerImpl.java#L289)):*
<details><summary><i>Previously, mtran7 wrote…</i></summary><blockquote>

I don't want to try and use the split method on an undefined or null variable. Please let me know if there's a better way to handle.
</blockquote></details>

yes the better way to do it is `StringUtils.isNotBlank` you can import `import static org.apache.commons.lang3.StringUtils.isNotBlank;` to use it.

**Review by Xin-Harness — APPROVED** [review]
> Reviewed 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45798)* status: 3 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by neelam-harness — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r8, 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45798#-:-NS8DwvSDTK9KjvXKImk:bnfp4nl)* status: 1. unresolved discussions.

**Pull request merged**